infra/gcp/lib_iam.sh

```diff
@@ -184,32 +235,139 @@ function ensure_removed_project_role_binding() {
     _ensure_removed_resource_role_binding "projects" "${project}" "${principal}" "${role}"
 }
 
-# Ensure that IAM binding has been removed from organization
+# Ensure that custom IAM role exists in scope and in sync with definition in file
 # Arguments:
-#   $1:  The principal (e.g. "group:k8s-infra-foo@kubernetes.io")
-#   $2:  The role name (e.g. "roles/foo.bar")
-function ensure_removed_org_role_binding() {
-    if [ ! $# -eq 2 -o -z "$1" -o -z "$2" ]; then
-        echo "ensure_removed_org_role_binding(principal, role) requires 2 arguments" >&2
+#   $1:  The scope of the role (e.g. "organization", "project")
+#   $2:  The id of the scope (e.g. "12345819", "k8s-infra-foo")
+#   $3:  The role name (e.g. "foo.barrer")
+#   $4:  The file (e.g. "/path/to/file.yaml")
+function _ensure_custom_iam_role_from_file() {
+    if [ ! $# -eq 4 -o -z "$1" -o -z "$2" -o -z "$3" -o -z "$4" ]; then
+        echo "_ensure_custom_iam_role_from_file(scope, id, name, file) requires 4 arguments" >&2
         return 1
     fi
 
-    local organization="${GCP_ORG}"
-    local principal="${1}"
-    local role="${2}"
+    local scope="${1}"
+    local id="${2}"
+    local name="${3}"
+    local file="${4}"
+    case "${scope}" in
+      organization | project ) ;;
+      * )
+        echo >/dev/stderr "_ensure_custom_iam_role_from_file(scope, id, name, file) scope must 'organization' or 'project'"
+        return 1
+        ;;
+    esac
 
-    _ensure_removed_resource_role_binding "organizations" "${organization}" "${principal}" "${role}"
+    local scope_flag="--${scope} ${id}"
+
+    local before="${tmp_dir}/custom-role.before.yaml"
+    local ready="${tmp_dir}/custom-role.ready.yaml"
+    local after="${tmp_dir}/custom-role.after.yaml"
+
+    # detect if we should create or update and dump role; silently ignore error
+    verb="update"
+    if ! (gcloud iam roles describe ${scope_flag} "${name}" | yq -Y 'del(.etag)' >"${before}") >/dev/null 2>&1; then
+        verb="create"
+    fi
+
+    # deleted roles can be undeleted within 7 days; after that must wait 30 days to create a role with same id
+    # ref: https://cloud.google.com/iam/docs/creating-custom-roles#deleting-custom-role
+    if <"${before}" grep -q "^deleted: true"; then
+        gcloud iam roles undelete ${scope_flag} "${name}"
+    fi
+
+    # name is foo.bar, but gcloud wants scopes/id/role/foo.bar in the file
+    local full_name="${scope}s/${id}/roles/${name}"
+    <"${file}" sed -e "s|^name: ${name}|name: ${full_name}|" >"${ready}"
+    gcloud iam roles "${verb}" ${scope_flag} "${name}" --file "${ready}" | yq -Y 'del(.etag)' > "${after}"
```

Comment:
is it kosher to depend on yq?  It's not installed on my desktop, for example :)  What package is it in?

No option for JSON?

Reply:
Oops, I saw the ensure.  NM